### PR TITLE
executor, statistics: drop the samples whose length is too long

### LINKED ISSUE #31726: analyze failure on news_item of tpc-e with 5k rows
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

restore tpc-e 5k customer dataset

### 2. What did you expect to see? (Required)
```
mysql> select count(*) from news_item;
+----------+
| count(*) |
+----------+
|     5000 |
+----------+
1 row in set (0.01 sec)
mysql> analyze table news_item;
ERROR 8004 (HY000): Transaction is too large, size: 104969237

CREATE TABLE `news_item` (
  `ni_id` bigint(12) NOT NULL,
  `ni_headline` varchar(80) NOT NULL,
  `ni_summary` varchar(255) NOT NULL,
  `ni_item` mediumblob NOT NULL,
  `ni_dts` datetime NOT NULL,
  `ni_source` varchar(30) NOT NULL,
  `ni_author` varchar(30) DEFAULT NULL,
  PRIMARY KEY (`ni_id`) /*T![clustered_index] CLUSTERED */
);
```
### 3. What did you see instead (Required)
analyze succeed.

### 4. What is your TiDB version? (Required)

master version
```
Release Version: v5.5.0-alpha
Edition: Community
Git Commit Hash: 8ee88265cdfdb9c1b8c6142cd8852628521551ce
Git Branch: heads/refs/tags/v5.5.0-alpha
UTC Build Time: 2022-01-13 14:57:06
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/32535, close https://github.com/pingcap/tidb/issues/31726
Problem Summary:

See the issue for more details.

### What is changed and how it works?

We dropped the large values before building the top-n and the histogram.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: executor/analyze.go [hunk 1]
```diff
@@ -1346,6 +1346,10 @@ workLoop:
 					continue
 				}
 				val := row.Columns[task.slicePos]
+				// If this value is to very big, we think that it not a value that can occur many times. So we don't record it.
```
Comment (xuyifangreeneyes) [inline executor/analyze.go:1349]:
```suggestion
				// If this value is very big, we think that it is not a value that can occur many times. So we don't record it.
```

File: executor/analyze.go [hunk 2]
```diff
@@ -1370,13 +1374,19 @@ workLoop:
...
+		indexSampleCollectLoop:
 			for _, row := range task.rootRowCollector.Base().Samples {
 				if len(idx.Columns) == 1 && row.Columns[idx.Columns[0].Offset].IsNull() {
 					continue
 				}
 
 				b := make([]byte, 0, 8)
 				for _, col := range idx.Columns {
+					// If the index value contains one value which is too long, we think that it's a value that doesn't occur many times.
+					if (col.Length != types.UnspecifiedLength && col.Length > statistics.MaxSampleValueLength) ||
```
Comment (xuyifangreeneyes) [inline executor/analyze.go:1386]: Why do we check `col.Length`? Even if `col.Length > statistics.MaxSampleValueLength`, the length of the sample value may still be small. Besides, we can move the check out of the sample loop.

**Comment by winoros** [comment]
> We can also close #31726 by the pr?

Yeah, let me add it.

**Review by chrysan — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**